databases/core.py

```diff
@@ -237,8 +237,12 @@ def __init__(self, backend: DatabaseBackend) -> None:
     async def __aenter__(self) -> "Connection":
         async with self._connection_lock:
             self._connection_counter += 1
```

Comment:
```suggestion
            if self._connection_counter == 0:
                await self._connection.acquire()
                self._connection_counter += 1
```

Reply:
If any exception happens it won't be incremented at all, so we don't have to decrement in the except part.

Reply:
@taybin @qwerty what do you think? This is simpler and won't need the extra test,

Reply:
Does this still work with concurrent requests?  It removes the semaphore.  In which case it doesn't even need the counter.  The connection lock would be enough.

But is the connection lock enough or do we need the semaphore?

Reply:
@taybin I'm not really sure, I'll try a test, if it doesn't work out, we'll merge it as is.